design-guidelines/form-ux.md

```diff
@@ -0,0 +1,189 @@
+
+The context in which a form is rendered depends on the complexity of the form and how critical the form's purpose is to the rest of the app.
+
+### Full page
+
+A form may be rendered on it's own page helps users to focus on completing their task with minimal distractions in the UI.
+
+Render a form on it's own page if the form:
+
+- is complex and has many fields
+- will be used to complete an action that is very important
+
+### Section of a page
+
+A form may be rendered on a page alongside additional content to avoid taking users out of their current flow.
+
+Render a form alongside other content if:
+
+- the user may need to reference page content to help them fill out the form correctly
+- the form is not so complex that it deserves it's own page
+- it would be too disruptive to take a user away from the current page
+- the page has content that is related to the form
+
+### Dialog modal
+
+A form may be rendered in a dialog modal to keep the user on their current page while directing their focus to a form. The first input of the form should receive focus when the modal is opened.
+
+Render a form in a dialog modal if:
+
+- there are very few fields in the form
+- there is no logical place to render the form on the current page
+- the user won't need to reference content that will be covered by the dialog
+
+### Inline inputs
+
+In lieu of rendering an entire form, a single input can be used in place of the data it controls. Inputs would replace the data once the user has explicity toggled edit mode.
+
+Use an inline input if it would be too disruptive to show an entire form or take the user to a new page.
+
+## Structure and hierarchy
+
+Forms are designed with a consistent hierarchy and structure to make it easier for users to scan the inputs and submit their data.
+
+Every form will have at least one section, but form sections should always be used when a complex form can be broken down into distinct sections of related fields.
+
+For more information about the parts that make up a form, see the Storybook documentation for each component under the "Form structure" section.
+
+### Field order
+
+Order fields by their relative level of importance. The inputs that are most critical for the user to complete their task should come first.
+
+### Field grouping
+
+Fields that are closely related should be visually grouped together.
+
+For more information on grouped fields, see the Storybook documentation for the `FieldList` and `FieldGroup` components.
+
+### Form actions
+
+A form should contain at least two buttons: one to submit the form, and one to navigate away from the form without submitting any of the input.
+
+These buttons should be highly visible, and easy for the user to press.
+
+### Multi-part forms
+
+Forms should be kept as succinct as possible, but some forms require a lot of data input. Break complex forms into multiple forms when it will help the user focus on what data they're inputting.
+
+#### Tabbed form pattern
```

Comment:
i just realized that i like the navigability of https://getbootstrap.com/docs/4.0/components/forms/ with that scrollSpy-index on the right. that pattern also makes it rather simple to provide a good experience on mobile.

at first sight it looks almost the same as our tabbed form pattern except for the form being one big form. i feel like that should also makes implementation less complex sometimes. i also realize that it goes against "showing too many fields at once" - but hey, we're in a complex form anyways, so some tradeoff is inevitable here. 

this is just an idea as i personally like the scrolling experience better (it even allows me to `cmd+f` for a field name, which the tabbed form does not. but that again is a personal habit). so i'm not proposing a change here but rather want to know if an approach like this has ever been discussed.

thanks for reading and sorry for opening that kind of bottle :)

Reply:
I agree with you - the "scrollspy" pattern is the best of both worlds.

I fully support updating this part of the guidelines if the designers agree :)

Reply:
scrollspy is great for reading like docs  (where you saw this component in use) but not necessarily best for forms. the advantage of our current vertical tab model is it allows us to have dependent steps moving forward and restricting those more (wizard like) and even now, it creates hard (yet not enforced) steps users are encouraged to go though. 
long scrolling can feel overwhelming when not properly used

I think it's a great component for view-only pages, especially text heavy ones. I like it and I would very much welcome it, but not necessarily for create/edit forms. If you see such cases of such use in the wild, please share so we can review together.

what could be very cool for forms is to have the typeforms experience for some cases where good fit which is best of both worlds of single page yet appropriate sized chunks.
https://www.typeform.com/product/?tf_campaign=brand_8653539325_v2&tf_source=google&tf_medium=paid&tf_content=87781712620_407037804926&tf_term=typeform&tf_dv=c&tf_matchtype=e&tf_adposition=&gclid=EAIaIQobChMIusbdtdjN6wIVlh6tBh2MnwDtEAAYASAAEgL_G_D_BwE